### ISSUE TITLE
Cobalt adaptor - 'total_cpu_count' job description parameter may be irrelevant/unneeded

### ISSUE THREAD

**Issue by itomaldonado** [radical-cybertools/saga-python#620]
Creating an issue more for discussion than anything else, since I'd like to get some input from you Andre~

Currently the Cobalt adaptor has these three job description parameters to control the number of nodes, cpus, mpi ranks.

```
# Example, 16-ranks-per-host, total of 8192 ranks, equate 8192 cpus (1 per rank)
jd.processes_per_host   = 16
jd.number_of_processes  = 8192
jd.total_cpu_count      = 8192 # 16 cores per node == 512 nodes
```

Here, we want to run a task in *512* nodes at 16-ranks per node and 1 rank per core. Cobalt (specifically *Mira*) has a static cores-per-node count of `16`.

The original idea was to use the `jd.total_cpu_count` parameter since you can have multiple ranks per core (up to 4 ranks per core), so it would make node calculations easy. But after some thinking I believe we may be able to now have it and just rely on `jd.processes_per_host` and `jd.number_of_processes` since with these, the number of nodes can be extracted (with some assumptions~)

Here is what I mean:

### Example 1: Single rank per node
* Ranks-per-node: 1
* Total Number of Ranks: 512

Here, since we have only one rank per node and a total of 512 ranks, we will use 512 nodes and since the number of cores is static in *Mira* (`16 cores-per-node` as per documentation) we can assume we need a total of 8192 cores.

We get that:
`cores = 16 * (jd.number_of_processes / jd.processes_per_host)`

### Example 2: Single rank per core
* Ranks-per-node: 16
* Total Number of Ranks: 8192

Similar as above with the slight difference that now all each rank will use up a core (instead of a rank using all cores in a node), but the calculations use the same formula.

`cores = 16 * ( 8192 / 16 ) = 8192`

### Example 3: Multiple ranks per core
* Ranks-per-node: 64
* Total Number of Ranks: 32,768

This example is also similar, but now we have 4 ranks per core (this is the upper limit in *Mira*). The calculations use the same formula.

`cores = 16 * ( 32,768 / 64 ) = 8192`

### Example 4: Multiple ranks per core, not full usage
* Ranks-per-node: 64
* Total Number of Ranks: 32,705

This example is different, now the total number of ranks do not use up all available space, which actually breaks our previous formula.

`cores = 16 * ( round_up(32,705 / 64) ) = 511 # should be 512 !!!`

This is were documentation helps again, in *Mira* you cannot request *partial nodes* so in cases where part of the node is needed, we can use the following formula (which can also be generalized).

`cores = 16 * ( ceiling[32,705 / 64] ) = 512 # Correct!`

or in python 2.7~

```
import math
cores = int( 16 * math.ceil( float( 32705 ) / float( 64 ) ) ) # = 512
```

### In short

We can drop the parameter `jd.total_cpu_count` and calculate it with the other two since we have the following assumptions for *Mira Blue Gene/Q*:
* Cores per node: 16
* Max ranks per node: 64
* Nodes cannot be partially scheduled 

Finally, we can also provide defaults for the other two parameters (as *Mira* does).
```
# Defaults
jd.processes_per_host   = 1
jd.number_of_processes  = ( jd.processes_per_host )

```

Source: [Mira Documentation](http://www.alcf.anl.gov/user-guides/running-jobs#submitting-a-script-job)

**Comment by iparask** [comment]
This is also an interesting issue and remains open

**Comment by iparask** [comment]
Check all adaptors and make sure they ceil the division of total number of CPUs with CPUs per node.
